zipline/pipeline/loaders/blaze/core.py

```diff
@@ -729,25 +729,36 @@ def __new__(cls,
             expr,
             deltas,
             checkpoints,
-            odo_kwargs or {},
+            frozenset((odo_kwargs or {}).items()),
         )
 
     def __repr__(self):
         # If the expressions have _resources() then the repr will
         # drive computation so we take the str here.
-        cls = type(self)
-        return super(ExprData, cls).__repr__(cls(
+        return repr(_expr_data_base(
             str(self.expr),
             str(self.deltas),
             str(self.checkpoints),
-            self.odo_kwargs,
+            dict(self.odo_kwargs),
         ))
 
-    def __hash__(self):
-        return id(self)
+    @staticmethod
+    def _expr_eq(a, b):
+        return a is b is None or a.isidentical(b)
 
     def __eq__(self, other):
-        return self is other
+        if not isinstance(other, ExprData):
+            return NotImplemented
+
+        return (
+            self.expr.isidentical(other.expr) and
+            self._expr_eq(self.deltas, other.deltas) and
+            self._expr_eq(self.checkpoints, other.checkpoints) and
+            self.odo_kwargs == other.odo_kwargs
```

Comment:
Any worry that `odo_kwargs` might contain object that aren't comparable and/or objects that don't have the equality semantics we want?

Reply:
No, these are mostly simple objects like strings, bytes, or ints. I think the only real use site we have for this is `bytes`.